src/test/validation_block_tests.cpp

```diff
@@ -0,0 +1,179 @@
+    pblock->nTime = ++time;
+
+    CMutableTransaction txCoinbase(*pblock->vtx[0]);
+    txCoinbase.vout.resize(1);
+    txCoinbase.vin[0].scriptWitness.SetNull();
+    pblock->vtx[0] = MakeTransactionRef(std::move(txCoinbase));
+
+    return pblock;
+}
+
+std::shared_ptr<CBlock> FinalizeBlock(std::shared_ptr<CBlock> pblock)
+{
+    pblock->hashMerkleRoot = BlockMerkleRoot(*pblock);
+
+    while (!CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
+        ++(pblock->nNonce);
+    }
+
+    return pblock;
+}
+
+// construct a valid block
+const std::shared_ptr<const CBlock> GoodBlock(const uint256& prev_hash)
+{
+    return FinalizeBlock(Block(prev_hash));
+}
+
+// construct an invalid block (but with a valid header)
+const std::shared_ptr<const CBlock> BadBlock(const uint256& prev_hash)
+{
+    auto pblock = Block(prev_hash);
+
+    CMutableTransaction coinbase_spend;
+    coinbase_spend.vin.push_back(CTxIn(COutPoint(pblock->vtx[0]->GetHash(), 0), CScript(), 0));
+    coinbase_spend.vout.push_back(pblock->vtx[0]->vout[0]);
+
+    CTransactionRef tx = MakeTransactionRef(coinbase_spend);
+    pblock->vtx.push_back(tx);
+
+    auto ret = FinalizeBlock(pblock);
+    return ret;
+}
+
+void BuildChain(const uint256& root, int height, const unsigned int invalid_rate, const unsigned int branch_rate, const unsigned int max_size, std::vector<std::shared_ptr<const CBlock>>& blocks)
+{
+    if (height <= 0 || blocks.size() >= max_size) return;
+
+    bool gen_invalid = GetRand(100) < invalid_rate;
+    bool gen_fork = GetRand(100) < branch_rate;
+
+    const std::shared_ptr<const CBlock> pblock = gen_invalid ? BadBlock(root) : GoodBlock(root);
+    blocks.push_back(pblock);
+    if (!gen_invalid) {
+        BuildChain(pblock->GetHash(), height - 1, invalid_rate, branch_rate, max_size, blocks);
+    }
+
+    if (gen_fork) {
+        blocks.push_back(GoodBlock(root));
+        BuildChain(blocks.back()->GetHash(), height - 1, invalid_rate, branch_rate, max_size, blocks);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(processnewblock_signals_ordering)
+{
+    // build a large-ish chain that's likely to have some forks
+    std::vector<std::shared_ptr<const CBlock>> blocks;
+    while (blocks.size() < 50) {
+        blocks.clear();
+        BuildChain(Params().GenesisBlock().GetHash(), 100, 15, 10, 500, blocks);
+    }
+
+    bool ignored;
+    CValidationState state;
+    std::vector<CBlockHeader> headers;
+    std::transform(blocks.begin(), blocks.end(), std::back_inserter(headers), [](std::shared_ptr<const CBlock> b) { return b->GetBlockHeader(); });
+
+    // Process all the headers so we understand the toplogy of the chain
+    BOOST_CHECK(ProcessNewBlockHeaders(headers, state, Params()));
+    ProcessNewBlock(Params(), std::make_shared<CBlock>(Params().GenesisBlock()), true, &ignored);
+
+    // subscribe to events (this subscriber will validate event ordering)
+    TestSubscriber sub(chainActive.Tip()->GetBlockHash());
```

Comment:
As discussed on IRC, there's a race condition where callbacks scheduled prior to a client subscribing to callbacks can be delivered to those clients, if the scheduler didn't finish draining its queue before the new subscriber is added.

That causes this test to rarely fail, because we initialize the TestSubscriber with the genesis block, even though it might be notified of the genesis block's connection (which has already happened at this point!) if the callbacks were delayed.  This can be reproduced by adding a `MilliSleep()` call to `SingleThreadedSchedulerClient::ProcessQueue()`.

The call to `ProcessNewBlock` at line 141 above doesn't fix this issue; instead we need to ensure that there are no outstanding callbacks at the time the TestSubscriber is created, for instance by calling `SyncWithValidationInterfaceQueue()`.  Also we should hold `cs_main` before accessing `chainActive`:

```
    SyncWithValidationInterfaceQueue();

    const CBlockIndex *initial_tip = nullptr;
    {
        LOCK(cs_main);
        initial_tip = chainActive.Tip();
    }
```